scarlet2/nn.py

```diff
@@ -0,0 +1,156 @@
+def calc_grad(x, trained_model):
+    """Calculates the gradient of the log-probability using the ScoreNet model"""
+    # perform padding if needed
+    t = 0.0 # corresponds to noise free gradient
+    x = jnp.float32(x) # cast to float32
+    x, ScoreNet, pad_lo, pad_hi, pad = pad_fwd(x, trained_model)
+    assert (x.shape[1] % 32) == 0, f"image size must be 32 or 64, got: {x.shape[1]}"
+    if jnp.ndim(x) == 2:
+        x = jnp.expand_dims(x, axis=0)
+        nn_grad = ScoreNet(x, t=t)
+        nn_grad = jnp.squeeze(nn_grad, axis=0)
+    else:
+        nn_grad = ScoreNet(x, t=t)
+    # return to original size
+    if pad: nn_grad = pad_back(nn_grad, pad_lo, pad_hi)
+    return nn_grad
+
+
+# ------------------------------------------------- #
+#   transformation corrections in training space    #
+# ------------------------------------------------- #
+# jax gradient function
+def vgrad(f, x):
+    y, vjp_fn = vjp(f, x)
+    return vjp_fn(jnp.ones(y.shape))[0]
+
+# transform to training space
+def transform(x):
+    sigma_y = 0.10
+    return jnp.log(x + 1) / sigma_y
+# ------------------------------------------------- #
+
+
+# inheritance from Distribution class
+class NNPrior(dist.Distribution):
+    support = constraints.real_vector
+
+    """Prior distribution based on a neural network"""
+
+    # construct custom vector-jacobian product
+
+    def __init__(self, trained_model='None', shape=1, validate_args=None):
+        # TODO: what's up with these globals?
+        # NOTE: this is a hack to get around the fact that    
+        # jax doesn't allow flexible arguments to be passed   
+        # to the log_prob function                           
+        global custom_model
+        custom_model = trained_model
+
+        # TODO: needs shape of the model images
+        event_shape = jnp.shape(shape)
+        super().__init__(
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+    def sample(self, key, sample_shape=()):
+        # TODO: add ability to draw samples from the prior, if desired
+        # NOTE: this will have a fixed size, ie the training data array size
+        raise NotImplementedError
+
+    def mean(self):
+        raise NotImplementedError
+    
+    @custom_vjp
+    def log_prob(x):
+        return 0.0
+
+    def log_prob_fwd(x):
+        # Returns primal output and residuals to be used in backward pass by f_bwd
+        x = transform(x) # transform to trainign space
+        nn_grad = calc_grad(x, custom_model)
+        nn_grad = vgrad(transform, x) * nn_grad # chain rule
+        return 0.0, nn_grad  # cannot directly call log_prob in Class object
+
+    def log_prob_bwd(res, g):
+        nn_grad = res # Get residuals computed in f_fwd
+        vjp = (g * nn_grad,) # create the vector (g) jacobian (nn_grad) product
+        return vjp
+    
+    # register the custom vjp 
+    log_prob.defvjp(log_prob_fwd, log_prob_bwd)
```

Comment:
I think that can be done cleaner, i.e. without a global variable. See [here](https://github.com/google/jax/discussions/16540#discussioncomment-6265352).

I mean something like this (all before of the class context)
```
from functools import partial
@partial(jax.custom_vjp, nondiff_argnums=(0,))
def _log_prob(model, x):
    return 0
    
def log_prob_fwd(model, x):
    x = transform(x)
    nn_grad = calc_grad(x, model)
    ...
```
and then inside of the prior class, you should be able to do this:
```
def log_prob(self, x):
    return _log_prob(self, x)
```
It might even work to pull the partial trick _inside_ of the class definition, which means that you don't have to effectively duplicate the `log_prob` method.

Reply:
This is all done here (https://github.com/pmelchior/scarlet2/blob/matt-nn-branch/scarlet2/nn.py) will update pull request. Sample updated runscript here (https://github.com/pmelchior/scarlet2/blob/matt-nn-branch/sample_runscript.ipynb)

Note, the training data transform can now be specified by the user when initialising the prior then the Jacobian transform to convert to correct gradients is done with jax.